app/src/main/java/com/anysoftkeyboard/gesturetyping/GestureTypingDetector.java

```diff
@@ -117,37 +127,33 @@ public void destroy() {
                 .observeOn(RxSchedulers.mainThread());
     }
 
-    private static int[] generatePath(char[] word, Iterable<Keyboard.Key> keysList, int curvatureSize, WorkspaceData workspaceData) {
+    private static int[] generatePath(char[] word, SparseArray<Keyboard.Key> keysByCharacter, WorkspaceData workspaceData) {
         workspaceData.reset();
         //word = Normalizer.normalize(word, Normalizer.Form.NFD);
         char lastLetter = '\0';
 
         // Add points for each key
         for (char c : word) {
-            c = Dictionary.toLowerCase(c);
+            c = Character.toLowerCase(c);
             if (lastLetter == c) continue; //Avoid duplicate letters
 
-            Keyboard.Key keyHit = null;
-            outer:
-            for (Keyboard.Key key : keysList) {
-                for (int i = 0; i < key.getCodesCount(); ++i) {
-                    if (Dictionary.toLowerCase((char) key.getCodeAtIndex(i, false)) == c) {
-                        keyHit = key;
-                        break outer;
-                    }
-                }
-            }
+            Keyboard.Key keyHit = keysByCharacter.get(c);
 
             if (keyHit == null) {
-                Logger.w(TAG, "Key %s not found on keyboard!", c);
-                continue;
+                // Try finding the base character instead, e.g., the "e" key instead of "é"
+                char baseCharacter = Dictionary.toLowerCase(c);
```

Comment:
I see that here you used `Dictionary. toLowerCase` instead of the `Character` counterpart, why?

Reply:
`Character.toLowerCase` gives you the actual lowercase character, e.g. `É` --> `é`. But `Dictionary.toLowerCase` gives you the lowercase base character, e.g., `É` --> `e`. Previously, the gesture typing was always using `Dictionary.toLowerCase`, which meant there was no difference between, say, `ä` and `a`. However, for keyboards, where both characters are present as different keys (e.g., the German keyboard), we want a different behavior in the path generation.
If a key for `ä` is present on the keyboard, we assume the user would swipe over it to use it, and not the character `a`. This allows for him to type words like `Männer` properly. Therefore we look for the `Character.toLowerCase`.
But if they want to type `café` on a German keyboard, there is no key for `é`. Instead, we would assume they would swipe over `e`. Therefore we look for the `Dictionary.toLowerCase` if we couldn't find the `Character.toLowerCase`.
This is why we should first look for a key which matches the lowercase exactly, and if we don't find such a key, we try looking for its base character.